pkg/icingaredis/client.go

```diff
@@ -208,28 +208,22 @@ func (c *Client) StreamLastId(ctx context.Context, stream string) (string, error
 // Each call blocks at most for the duration specified in Options.BlockTimeout until data
 // is available before it times out and the next call is made.
 // This also means that an already set block timeout is overridden.
-// For each of the special $ IDs, if the first call fails, the last stream ID is retrieved and used,
-// so that entries added between calls are not lost.
 func (c *Client) XReadUntilResult(ctx context.Context, a *redis.XReadArgs) ([]redis.XStream, error) {
-	// Replace $ with the latest stream IDs, so we don't miss any entries added between calls.
-	middle := len(a.Streams) / 2
-	for i := middle; i < len(a.Streams); i++ {
-		if a.Streams[i] == "$" {
-			lastId, err := c.StreamLastId(ctx, a.Streams[i-middle])
-			if err != nil {
-				return nil, errors.Wrap(err, "can't get last stream ID of "+a.Streams[i-middle])
-			}
-			a.Streams[i] = lastId
-		}
-	}
-
 	a.Block = c.Options.BlockTimeout
 
+	XRead := func() *redis.XStreamSliceCmd {
+		// Don't consume go-redis retries if the result is not available in time anyway.
+		ctx, cancelCtx := context.WithTimeout(ctx, a.Block)
+		defer cancelCtx()
+
+		return c.XRead(ctx, a)
+	}
```

Comment:
Doesn’t this (due to the context timeout is equal to, but starts just before the Redis server-side one) just kicks you out of XRead with context.DeadlineExceeded just before hitting redis.Nil? No, wait, on failure it s/$ERR/context.DeadlineExceeded/ after 1s and doesn’t retry anymore, but you do in the loop, right?

Reply:
Yes that could happen.